### DISCUSSION TITLE
Ошибки при кастомной обработки при импорте CSV

### DISCUSSION THREAD

**Discussion by mgmman** [totumonline/totum-ru-issues-and-discussions#423] [ОШБК!] [answered]
при попытки использовать в кастомной предобработке импортируемого CSV функции setList totum выдает UnexpectedError и трейсбак на php.
При замене setList на insertList (больше ничего не менялось) все работает как положено, однако, возникает ошибка декодировки русских букв (смена кодировки загружаемого файла не решает проблему).
Стоит также отметить, что при использовании встроенной функции "CSV-импорт" для загрузки того же файла все работает нормально и проблем с декодировкой не возникает
![Снимок экрана (1)](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/115487609/f8332c62-7312-4b0b-a865-189d0ec18d0a)
![Снимок экрана (2)](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/115487609/462aba9b-a302-4680-a9f1-df767209d219)
![Annotation 2023-08-02 110002](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/115487609/046299d0-e3b6-4b43-a23f-0dbf69f0f07a)
![Annotation 2023-08-02 110042](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/115487609/9d9db0ee-23dd-4778-8ff5-24da4c9b9b56)
![Annotation 2023-08-02 133542](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/115487609/b8d83746-6b82-446a-9172-c98979120e44)

**Comment by ttm-alexeypolunin** [comment]
По кодировке ничего не скажу, а по ошибке - проверьте файл поиском на переносы строк в неожиданных местах и на `;` в значениях. Если вы его выгружаете из excel то оба этих варианта могут быть тк CSV в этом случае является экстра-ненадежным вариантом.

Для того, что бы с большей надежностью читать xlsx — сервер Сервисов. Функция serviceXlsxParser(file: ; withformats: )

+ **file** — сюда путь к файлу xlsx на сервере, получить его можно #file_field[0][file]

+ **withformats** — можете не использовать, он возвращает информацию по форматированию ячеек

**Reply by ttm-alexeypolunin** [reply]
А почему setList — он устанавливает значения в конкретные строки, а вам вроде как надо добавить новые строки, но я может чего не понимаю.

Второй момент, который виден из скрина, что вы несколько раз вызываете split_in_row — зафиксируйте ее, тк сейчас идет расчет при каждом обращении

```
~split_in_row: ...
```

**Reply by mgmman** [reply]
Файл который я пытался обработать был экспортирован в CSV из самого тотума с помощью встроенной функциональности. Так что проблема навряд ли заключается в этом.

**Reply by ttm-alexeypolunin** [reply, marked as answer]
У вас странное с setList — эта функция изменения значений полей уже существующих строк. Если у вы изменяете строки, то надо указать where.

Если добавляете, то надо использовать insert.

setList еще ставит одинаковые значения во все строки попадающее под условие. Если надо вставить разное в разные строки, то надо использовать setListExtended
